web/src/narrow.js

```diff
@@ -405,93 +421,135 @@ export function activate(raw_terms, opts) {
+        }
+
+        narrow_state.set_has_shown_message_list_view();
+        // Show the new set of messages. It is important to set message_lists.current to
         // the view right as it's being shown, because we rely on message_lists.current
         // being shown for deciding when to condense messages.
+        // From here on down, any calls to the narrow_state API will
+        // reflect the requested narrow.
+
         message_lists.update_current_message_list(msg_list);
 
+        let select_immediately = true;
         let then_select_offset;
-        if (id_info.target_id === id_info.final_select_id) {
-            then_select_offset = opts.then_select_offset;
-        }
-
-        const select_immediately = id_info.local_select_id !== undefined;
-
-        {
-            let anchor;
-
-            // Either we're trying to center the narrow around a
-            // particular message ID (which could be max_int), or we're
-            // asking the server to figure out for us what the first
-            // unread message is, and center the narrow around that.
-            switch (id_info.final_select_id) {
-                case undefined:
-                    anchor = "first_unread";
-                    break;
-                case -1:
-                    // This case should never happen in this code path; it's
-                    // here in case we choose to extract this as an
-                    // independent reusable function.
-                    anchor = "oldest";
-                    break;
-                case LARGER_THAN_MAX_MESSAGE_ID:
-                    anchor = "newest";
-                    break;
-                default:
-                    anchor = id_info.final_select_id;
+        let select_opts = {
+            empty_ok: false,
+            force_rerender: true,
+        };
+        if (filter.is_in_home()) {
+            select_opts = {
+                empty_ok: true,
+                force_rerender: false,
+            };
+            if (unread.messages_read_in_narrow) {
+                // We read some unread messages in a narrow. Instead of going back to
+                // where we were before the narrow, go to our first unread message (or
+                // the bottom of the feed, if there are no unread messages).
+                id_info.final_select_id = message_lists.current.first_unread_message_id();
+            } else {
+                // We narrowed, but only backwards in time (ie no unread were read). Try
+                // to go back to exactly where we were before narrowing.
+                // We scroll the user back to exactly the offset from the selected
+                // message that they were at the time that they narrowed.
+                // TODO: Make this correctly handle the case of resizing while narrowed.
+                then_select_offset = message_lists.current.pre_narrow_offset;
+                id_info.final_select_id = message_lists.current.selected_id();
```

Comment:
This block seems to assume `message_lists.current.selected_id` and `message_lists.current.pre_narrow_offset` will be both `undefined` in the event that we hadn't visited that view yet or it otherwise wouldn't be valid to read these values ... I think that's correct, but it took me a while to confirm that.